Do not allow Sampler to allocate more than maxDoc size, better CB accounting

polyfractal opened the pull request:
The `sampler` agg creates a BestDocsDeferringCollector, which internally initializes a priority queue of size `shardSize`.  This queue is populated with empty `Object` sentinels, which is roughly 16b per object.  Similarly, the Diversified samplers createa DiversifiedTopDocsCollectors which internally track PQ slots with `ScoreDocKeys`, weighing in around 28kb

If the user sets a very abusive `shard_size`, this could easily OOM a node or cluster since these PQ are allocated up-front without any checks.

This PR makes sure that when we create the collector, it cannot be larger than the maxDoc so that we don't accidentally blow up the node. This is done in the ctor so the caller can't forget to run the check.

For good measure, this also adds in some CB accounting to try and track memory usage.

Finally, a redundant array creation is removed to reduce a bit of temporary memory.

Closes #34269

Commit:
Do not allow Sampler to allocate more than maxDoc size, better CB

The `sampler` agg creates a BestDocsDeferringCollector, which internally
initializes a priority queue of size `shardSize`.  This queue is
populated with empty `Object` sentinels, which is roughly 16b per
object.

Similarly, the Diversified samplers createa DiversifiedTopDocsCollectors
which internally track PQ slots with ScoreDocKeys, weighing in around
28kb

If the user sets a very abusive `shard_size`, this could easily OOM
a node or cluster since these PQ are allocated up-front without
any checks.

This commit makes sure that when we create the collector, it
cannot be larger than the maxDoc so that we don't accidentally blow
up the node.  We ensure the size is not greater than the overall
index maxDoc. This is done in the ctor so that a caller can't forget
to run the check.

For good measure, this also adds in some CB accounting to try and track
memory usage.

Finally, a redundant array creation is removed to reduce a bit of
temporary memory.

Commit:
Add tests

elasticmachine commented:
Pinging @elastic/es-analytics-geo

markharwood reviewed (commented):
LGTM but there's another PriorityQueue that needs protection/test coverage.
The Diversified sampler has a top-level PQ based on terms (e.g. website names) and for each term there's a PQ of the top `maxDocsPerValue` docs collectable for that unique term. Setting `maxDocsPerValue` to max int will also cause OOMs. This too should be capped to maxDoc but given it's a multiple of the top-level PQ size maybe there should be a soft limit? Could be tough figuring out what that would be but I guess adding the request-level circuit breaker would offer the required protection.
Thanks for looking into this.

polyfractal commented:
Ah cool, thanks for the note.  Will add some more protections in that area.  :+1:

Commit:
Review cleanup

- Add try/finally around CB
- Add maxDoc protection to diversified sampler's maxDocsPerValue
- Better accounting with RamUsageEstimator

Commit:
checkstyle

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/sampler/SamplerAggregator.java:
```diff
@@ -56,6 +58,8 @@
     public static final ParseField MAX_DOCS_PER_VALUE_FIELD = new ParseField("max_docs_per_value");
     public static final ParseField EXECUTION_HINT_FIELD = new ParseField("execution_hint");
 
+    static final long SCOREDOCKEY_SIZE = RamUsageEstimator.shallowSizeOfInstance(DiversifiedTopDocsCollector.ScoreDocKey.class);
```
@markharwood I went ahead and evaluated the full object using `shallowSizeOfInstance()` instead of stringing together constants.  I know it's a bit slower (reflection etc) but thought it might be ok since the value is a static final?

I don't have much experience with `RamUsageEstimator` though, so happy to switch back to adding together constants if that's better.

markharwood replied:
I don't have experience with `shallowSizeOfInstance` either but I see it used in several places in core and use as static final makes sense. I know it has some stuff in there to account for compressed vs uncompressed references.

Commit:
Merge remote-tracking branch 'origin/master' into sampler_circuit_breaker

polyfractal commented:
Think this is ready for a final review when you have a minute @markharwood.  :)

As discussed on slack, I wasn't able to add CB accounting for diversified samplers because Lucene doesn't expose details about the internal priority queues.  But I did add the `maxDoc` protection to the `maxDocsPerValue` field, so hopefully that combined with `max_buckets` will help.

markharwood commented:
>But I did add the maxDoc protection to the maxDocsPerValue field, so hopefully that combined with max_buckets will help.

That makes sense. LGTM but the last thing I think is missing is test coverage for a ridiculous `maxDocsPerValue` size in `DiversifiedSamplerTests`.

Commit:
Add test for diversified, consolidate maxDoc treatment

polyfractal commented:
Glad you noticed that missing test, turns out there was another location that needed maxDoc protection.  `DiversifiedMapSamplerAggregator` allocates a `BytesRefHash` based on size and could OOM.

I managed to cleanup/consolidate the maxDoc treatment so that it was a little more centralized and less c/p in different locations too.

Thanks for the help and reviews!

Commit:
Merge remote-tracking branch 'origin/master' into sampler_circuit_breaker

polyfractal commented:
@elasticmachine run elasticsearch-ci/bwc

Commit:
Merge remote-tracking branch 'origin/master' into sampler_circuit_breaker
